docs/proposals/20251001-rag-auto-indexer.md

```diff
@@ -0,0 +1,653 @@
+### PDF Handling
+
+PDF files are detected by content-type or file extension. The AutoIndexer will extract text from PDFs using the `PyMuPDF` library to extract text from each page. If the extraction is not successful, the document is skipped and an error is logged. This approach maximizes compatibility with a wide range of PDF files and ensures that both text and tabular data are captured where possible.
+### RAG Document Handling
+
+Each document ingested by the AutoIndexer will be uploaded to the RAG engine with metadata that includes:
+- The AutoIndexer name
+- Source URL or file path - used for sequential updates
+- Depending on file extension, we will use Sentence or Code Splitting automatically.
+
+This metadata enables filtering on ingested documents for drift detection, and allows for efficient cleanup or re-indexing of documents when source data changes.
+#### Git Repository handling
+
+For Git data sources, the AutoIndexer will:
+
+1. Clone the repository locally using the configuration provided (repo URL, branch, commit, etc.)
+2. On the first run, index all files matching the configured paths and upload them to the RAG engine with appropriate metadata.
+3. On subsequent runs, use the last processed commit (from the status ConfigMap) to determine the commit range. Use the diff between the current and last run commit to identify added, updated, deleted, or renamed files between the last and current commit.
+4. For added/updated files: re-index and upload to the RAG engine. For deleted files: remove from the RAG index. For renamed files: treat as delete+add.
+5. After processing, update the ConfigMap with the new commit hash and document counts.
+
+This incremental approach ensures efficient updates and minimizes unnecessary reprocessing, while keeping the RAG index in sync with the source repository.
+
+### Run Identity and Result Reporting
+
+Each AutoIndexer run must be uniquely identifiable so that operators and automation can correlate the Job, its logs, the generated ConfigMap, and the CRD status. To achieve this, we introduce a **Run ID** (`RUN_ID`) and a structured **ConfigMap** result for each run.
+
+#### Run ID
+
+* **Definition:** A unique string generated per AutoIndexer run.
+
+* **Format:**
+  * Run ID: `<autoindexer-name>-<job-name>`
+
+* **Propagation:**
+
+  * The controller injects the Run ID into the Job/CronJob pod spec as an environment variable `RUN_ID` as well as on labels/annotations.
+  * The Job uses the Run ID in logging.
+  * The Job uses the Run ID in naming its result ConfigMap.
+
+* **Usage:**
+
+  * Operators can trace a run end-to-end with:
+    * `kubectl get job -l autoindexer.kaito.ai/run-id=<RUN_ID>`
+    * `kubectl logs job/<job-name>`
+    * `kubectl get configmap autoindexer-<autoindexer_name>`
+  * The controller records `status.lastRunID` for correlation.
+
+#### Result ConfigMap
```

Comment:
I am ok with having configmaps to store all the job result history.

1) We need to add history length field in the crd, we cannot grow the number of configmap indefinitely with cron job.
2) configmap should be just used for auditing/troubleshooting, the controller logic should not rely on configmap, and required info for next job execution should be recorded in the status of autoindexer CR. This means even if someone deletes the configmaps, the controller should still function.

Reply:
After discussing we decided to go for giving the jobs permissions to edit the AutoIndexer status directly. We can also keep some scheduled jobs through .spec.successfulJobsHistoryLimit and .spec.failedJobsHistoryLimit on the CronJob spec. This will allow us to remove the ConfigMap completely